model-integration/src/main/java/ai/vespa/triton/TritonOnnxModelLoader.java

```diff
@@ -0,0 +1,86 @@
+package ai.vespa.triton;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.logging.Logger;
+
+/**
+ * Implements reference counting for loading and unloading ONNX models in NVIDIA Triton Inference Server.
+ * This is to avoid loading models that are already loaded and unloading models that are still in use.
+ * 
+ * @author glebashnik
+ */
+class TritonOnnxModelLoader {
+    private static final Logger logger = Logger.getLogger(TritonOnnxModelLoader.class.getName());
+    
+    // Lock for loading, unloading and counting operations to avoid race conditions with multiple evaluators.
+    private final Object modelReferenceLock = new Object();
+    private final Map<String, Integer> modelReferenceCounts = new HashMap<>();
+
+    private final TritonOnnxClient client;
+    private final boolean isExplicitControlMode;
+
+    TritonOnnxModelLoader(TritonOnnxClient tritonClient, boolean isExplicitControlMode) {
+        Objects.requireNonNull(tritonClient);
+        this.client = tritonClient;
+        this.isExplicitControlMode = isExplicitControlMode;
+    }
+
+    TritonOnnxClient.ModelMetadata loadModel(String modelName) {
+        Objects.requireNonNull(modelName);
```

Comment:
Consider add fine level logging when incrementing the reference and potentially loading the model